### ISSUE TITLE
Build charts for homepage stats section

### ISSUE THREAD

**Issue by ryancreatescopy** [ethereum/ethereum-org-website#2319]
Dependency: #2124

- [ ] Confirm each data source provides time series info (& range they provide)
- [ ] Select charting library to use
- [ ] (depending on above) add toggles to view (e.g. all / 1yr / 90 / 30 day)
- [ ] Add background trendline to each datapoint:
![Image 2021-03-08 at 9 38 00 AM](https://user-images.githubusercontent.com/8097623/110359141-ff2e4380-7ff1-11eb-8d3c-2b57b7acd032.jpg)

**Comment by wackerow** [comment]
@ryancreatescopy @samajammin 

- [x] **DeFi Pulse** - all/1y/1m can each be pulled (though each requires pulling two sets of data, All, and also Lightning Network to deduce Ethereum TVL
end point: https://data-api.defipulse.com/api/v1/defipulse/api/GetHistory?api-key={{DeFi_Pulse_Key}}&period=1y

- [x] **CoinGecko** - allows time ranges to pull historical prices and market cap
end point example: https://api.coingecko.com/api/v3/coins/ethereum/market_chart/range?vs_currency=usd&from=1580251070&to=1611873485
(would just have to pass desired date ranges when requesting) 

- [x] **Coin Metrics** - we're already pulling in full array of Txn Count for every day since genesis

- [ ] Etherscan options for node count tbd

Tried some of these using `react-charts` and they seem to be populating just fine. Will play around with this more.

Worth some discussion as to what time intervals we want to offer. Do we want to just do 30-day charts, or 1-year charts for all? Offer toggling between a few options? Or more interactive?

**Comment by ryancreatescopy** [comment]
So I think we can start by just showing 30 days. Again, we're most interested in showing how the network is a living breathing things with lots of use. Therefore, most recent data is more important, rather than comparisons to 2016 or whatever. So I suggest we start with 30 days and then maybe we can experiment with allowing a user to toggle based on whether that's technically worth it (for example, if it's going to add X more calls, perhaps it's not worth it) – a discussion for down the road.


Do you have the chart experiments in a branch anywhere? Would love to see them.

**Comment by wackerow** [comment]
@ryancreatescopy Very rough draft locally, I'll add Etherscan and push it today

**Comment by wackerow** [comment]
@ryancreatescopy Sorry for delay on this, had an issue that set me back, but now working. Holding on etherscan data for the moment while we work out historical data api endpoint. 

https://github.com/ethereum/ethereum-org-website/tree/metric-charting

Pushed the draft branch to above link, which just shows price data being pulled from CoinGecko over the last month and plots it on a chart using `react-charts`. I am not partial to this library if you have others to suggest. Otherwise I'll dig into it more, pull the remainder of the data feeds (need to test the DeFi Pulse serverless function, and add Coin Metrics) and work on some styling.

**Comment by ryancreatescopy** [comment]
I can't preview or build this. Maybe create a PR for it that we can collab on.

**Comment by IliadisVictor** [comment]
I would really like to work on this~

**Comment by wackerow** [comment]
@IliadisVictor Thanks for volunteering! Just curious, did you have a particular charting library in mind?

@samajammin @ryancreatescopy What do you think of handing this one off given our current load?

**Comment by samajammin** [comment]
Sounds good to me! Thanks for volunteering @IliadisVictor. I welcome contributions on this - but can we use one of these libraries for the charting?
- https://github.com/recharts/recharts
- https://github.com/airbnb/visx

**Comment by samajammin** [comment]
BTW the folks at Etherscan have built out an endpoint for node time series data!
https://api.etherscan.io/api?module=stats&action=nodecounthistory&startdate=2021-04-01&enddate=2021-04-29&sort=asc&apikey=YourApiKeyToken

**Comment by samajammin** [comment]
@ryancreatescopy I think we should provide a bit more design direction here if @IliadisVictor is going to pick this up.

I'd also suggest we start with a v1 of this feature by just picking a time series (e.g. 30 days) vs. also building in a toggle.

**Comment by IliadisVictor** [comment]
I am more comfortable working with recharts  , some more specific guidelines on design and on what we should begin with would be very helpful. !

**Comment by wackerow** [comment]
@ryancreatescopy If you have time for a quick figma mock sometime in the next week that could be shared that would be great. @samajammin thoughts on using recharts? Not familiar personally, but could look into it.

**Comment by IliadisVictor** [comment]
Hello guys once again , it seems i need the API keys for the data providers ( Etherscan , DeFiPulse exc ) to continue with the charts .

**Comment by wackerow** [comment]
@IliadisVictor Hey! Apologies for the delays getting back to you. Honestly not sure if we can share these keys...  I'll defer to @samajammin

An alternative in the meantime is potentially utilizing a draft PR and view the preview netlify builds. I know it's not ideal by any means and makes it much more challenging to build this. Will get back to you soon.

**Comment by wackerow** [comment]
@IliadisVictor Actually, I was thinking that these were restricted endpoints for some reason which I don't believe they are. Just merged in #3125 which updates instructions on how to create your own keys and use them (they go in the `.env` file). Have you given that a try yet by chance?

**Comment by IliadisVictor** [comment]
I have managed to create keys , thats not my problem ,  but most of the data we want , are restricted for premium users only , and i can only make free keys that limit me .  I have been able to go around on all the api's  except etherscan to get the daily Transactions amount . I dont know if you guys have premium keys , for etherscan and i dont know what we can do if you dont . 

For the rest ,  the charts are somewhat ready , the only thing missing is the design touch . I do have experience with adobe xd , and could make some mockups , but from what i understand its Ryan's job and i wouldn't want to make something that he wouldn't agree on . @wackerow @samajammin  @ryancreatescopy 



![194425146_1191857721276455_8007189145790729218_n](https://user-images.githubusercontent.com/56890377/120835985-b85ad600-c56d-11eb-818a-d9d1ca39d6a8.png)

**Comment by ryancreatescopy** [comment]
@wackerow we do have a mockup: https://www.figma.com/proto/5CFYpOcdM3dqTtsnSe3jcw/Homepage-v2?page-id=787%3A1604&node-id=787%3A4008&viewport=-291%2C-5%2C0.4031561315059662&scaling=min-zoom

**Comment by wackerow** [comment]
@ryancreatescopy Awesome, thanks for posting that!

**Comment by IliadisVictor** [comment]
It's really ok i understand we all have our obligations  , the last problem to tackle would be the etherscan issue , free keys wont do the job like the other api's  , if you do have one i really need it , or another solution that could help  ( you can see in the picture above ) . If there is any other issue you need assistance i would gladly help !  Thanks for the feedback @wackerow  @ryancreatescopy

**Comment by Andrew-Sofos** [comment]
Hello,  @samajammin  @wackerow  I am  working this issue together with @IliadisVictor  and I want to inform you about our progress and the problem that we are facing. First of all here is a screenshot of what we have achieved locally: 
![Screenshot (96)](https://user-images.githubusercontent.com/56540744/122805763-2aebe580-d2d2-11eb-9d23-6b6dc6bc5f18.png) 
The buttons are perfectly functional they toggle and they change the appropriate chart. As far as our problem is concerned, we made a draft pr in order to test the netlify deploy and get the data for the transactions but the charts unfortunately doesn't show up. We think that the problem isn't caused by the css or the recharts library but it is something else that we haven't figured out yet. Do you have any idea of what may be the cause? 
Also here is the link for the netlify deploy:
[https://deploy-preview-3304--ethereumorg.netlify.app/en/](url)

**Comment by wackerow** [comment]
![image](https://user-images.githubusercontent.com/54227730/122858385-0f331000-d2cf-11eb-8bc7-d0739d731480.png)
![image](https://user-images.githubusercontent.com/54227730/122863788-4d80fd00-d2d8-11eb-8003-0b6ec73323a5.png)
![image](https://user-images.githubusercontent.com/54227730/122864708-f9771800-d2d9-11eb-9282-5c35126ddedc.png)

My guess is it's something to do with one of your axios calls. Getting a lot of errors in the console... We are already calling endpoints from a useEffect hook within the StatsBoxGrid component, could we not consolidate these to reduce the number of calls? I would try to mimic the format already laid out for fetching the data. Some of those may required serverless (lambda) functions which are executed through Netlify (which is where the deploy previews should come in handy)

cc: @samajammin

**Comment by wackerow** [comment]
@Andrew-Sofos Taking another look here (looking at #3332), I see we're connecting an onClick handler to triggering a fetch event. For the purposes of this page, I think we should only fetch the data once within the useEffect hook. 

This thread has gotten quite long, so I'm going to recap the endpoints here:

## Price History Endpoint:
> https://api.coingecko.com/api/v3/coins/ethereum/market_chart/range?vs_currency=usd&from=1623974772&to=1624493174
> (No API-key required)

### outputs:
```json
{
    "prices": [
        [
            1623978151025,
            2366.0070107318256
        ],
        [
            1623981797278,
            2343.142364202033
        ],
...trimmed_for_brevity...
        [
            1624489214516,
            1954.3110467228803
        ],
        [
            1624492822644,
            1971.1059779809414
        ]
    ],
    "market_caps": [
        [
            1623978151025,
            275527080555.823
        ],
        [
            1623981797278,
            272252927146.43356
        ],
...
        [
            1624489214516,
            227517901836.85532
        ],
        [
            1624492822644,
            229518675836.76538
        ]
    ],
    "total_volumes": [
        [
            1623978151025,
            25089648716.91813
        ],
        [
            1623981797278,
            24980124452.60261
        ],
...
        [
            1624489214516,
            28269281434.093475
        ],
        [
            1624492822644,
            28098744044.509525
        ]
    ]
}
```
_(Returns hourly results)_

## DeFI TVL Endpoint:
> https://data-api.defipulse.com/api/v1/defipulse/api/GetHistory?api-key={{DEFI_PULSE_API_KEY}}&period=3m&length=1

### outputs:
```json
[
    {
        "timestamp": "1624489200",
        "tvlUSD": 49638903489,
        "tvlETH": 25472829.829630014,
        "BTC": 162826.74604140996,
        "ETH": 8516847.740739772,
        "DAI": 2640845893.6191964
    },
    {
        "timestamp": "1624406400",
        "tvlUSD": 48438860331,
        "tvlETH": 25783729.00418384,
        "BTC": 150849.36302132002,
        "ETH": 8494427.831142772,
        "DAI": 2510971536.9099855
    },
    {
        "timestamp": "1624320000",
        "tvlUSD": 49308872901,
        "tvlETH": 26014061.367892906,
        "BTC": 153257.55564340003,
        "ETH": 8594570.700945776,
        "DAI": 2561255146.730863
    },
...
    {
        "timestamp": "1616889600",
        "tvlUSD": 44046384770,
        "tvlETH": 25710007.45388745,
        "BTC": 128723.58226603005,
        "ETH": 9260056.268671157,
        "DAI": 1552212900.4422352
    },
    {
        "timestamp": "1616803200",
        "tvlUSD": 44386858428,
        "tvlETH": 26123285.697974853,
        "BTC": 128447.07357252002,
        "ETH": 9213883.256491158,
        "DAI": 1560990430.3680203
    }
]
```

## Transaction Count Endpoint:
> https://api.etherscan.io/api?module=stats&action=dailytx&startdate=2021-06-18&enddate=2021-06-23&sort=asc&apikey={{ETHERSCAN_API_KEY}}
> (Note, can also use `desc` for sort to reverse order of results)

### outputs:
```json
{
    "status": "1",
    "message": "OK",
    "result": [
        {
            "UTCDate": "2021-06-18",
            "unixTimeStamp": "1623974400",
            "transactionCount": 1170182
        },
        {
            "UTCDate": "2021-06-19",
            "unixTimeStamp": "1624060800",
            "transactionCount": 1050189
        },
        {
            "UTCDate": "2021-06-20",
            "unixTimeStamp": "1624147200",
            "transactionCount": 1057288
        },
        {
            "UTCDate": "2021-06-21",
            "unixTimeStamp": "1624233600",
            "transactionCount": 1209567
        },
        {
            "UTCDate": "2021-06-22",
            "unixTimeStamp": "1624320000",
            "transactionCount": 1297644
        }
    ]
}
```

## Node Count Endpoint:
> https://api.etherscan.io/api?module=stats&action=nodecounthistory&startdate=2021-06-18&enddate=2021-06-23&sort=asc&apikey={{ETHERSCAN_API_KEY}}
> (Note, can also use `desc` for sort to reverse order of results)

### outputs:
```json
{
    "status": "1",
    "message": "OK",
    "result": [
        {
            "UTCDate": "2021-06-18",
            "TotalNodeCount": "5861"
        },
        {
            "UTCDate": "2021-06-19",
            "TotalNodeCount": "5774"
        },
        {
            "UTCDate": "2021-06-20",
            "TotalNodeCount": "5731"
        },
        {
            "UTCDate": "2021-06-21",
            "TotalNodeCount": "5818"
        },
        {
            "UTCDate": "2021-06-22",
            "TotalNodeCount": "5785"
        }
    ]
}
```

For dev/testing purposes, I would start with using these as dummy data to get the right data structure. If the charts are working with some dummy data, I would then investigate the axios calls currently being executed in the `useEffect` hook of this component. I wouldn't add more calls, but double-check / adjust the ones that are there to fetch the new endpoints with historical data where needed. 

I believe the CoinGecko API is the only one we simply call directly, and this can be updated to use the new endpoint (with market history rather than just current price).

The other three call serverless functions which then make the API request (found in `/src/lambda/`), which will also need to be updated with the new endpoints... these get trickier to test without API keys, but if things are set up correctly, they should load with the Netlify deploy preview (the one build for each PR).

Does that help at all?

**Comment by Andrew-Sofos** [comment]
Hello,  @wackerow  I did a deploy with only dummy values and no axios calls and I was expecting to see something like this that I have locally.
![Screenshot (99)](https://user-images.githubusercontent.com/56540744/123250496-c9ef2800-d4f2-11eb-9fea-71e39d6a0645.png)

Unfortunately, the charts didn't appear once again and I still get this error when I run the build.
![Screenshot (98)](https://user-images.githubusercontent.com/56540744/123250638-f73bd600-d4f2-11eb-9f2e-d5f67600b976.png)

So we can say that the problem doesn't  occur by an axios call but it is something else that blocks the charts from appearing. I suppose that I have successfully installed the recharts library, because if I haven't I would be getting some console errors which I don't. 
![Screenshot (100)](https://user-images.githubusercontent.com/56540744/123251294-b5f7f600-d4f3-11eb-9b57-315f2db95bb7.png)

Do you have any other idea of what may be the problem? 
[netlify deploy here](https://deploy-preview-3332--ethereumorg.netlify.app/en/)
[draft pr](https://github.com/ethereum/ethereum-org-website/pull/3332)

**Comment by Andrew-Sofos** [comment]
Hello again, @samajammin @wackerow I found the main reason why the charts aren't appearing it was some changes that I did on css in order to have responsive charts. They were perfect locally but on deploy some css actions like z-index for example aren't acceptable I suppose. I am doing some testing with new deploys in order to find out what is acceptable and what is not and I will come back with a new deploy and hopefully all the charts that we want.

**Comment by Andrew-Sofos** [comment]
Hello, @samajammin @wackerow  I found out why the charts weren't appearing but now I am facing another issue. I am getting these errors ![image](https://user-images.githubusercontent.com/56540744/123568877-e975a180-d7cd-11eb-8f61-b1f28d2f4804.jpeg) 

I have tried everything and I haven't found neither a solution nor the reason why they are appearing. Do you have any useful any idea of what may be the cause ?

**Comment by wackerow** [comment]
@Andrew-Sofos Thanks for the work on this, took a peek locally and looks like a great start! I'll try to help get you going again...

In your screenshot one is tracing back to `StatsBoxGrid.js` line 507 which is in the `etherscanNodesData` function... You have your API key card-coded in the URL 😄  _Please make sure to inactivate that API key_ on Etherscan now that it's been committed and exposed.

My guess is since some of these are refusing your API key, which is likely leaving one of the charting objects in your code as `undefined`, and then it's trying to call a key on `undefined`. 

We currently use lambda functions for some of these API calls to help avoid cross-origin (CORS) errors. These are in the folder `src/lambda/`. The current Etherscan calls that are performed are `etherscan.js` and `txs.js`. These use an internal API key that Netlify can access for previews/deploys, and we can access using `process.env.ETHERSCAN_API_KEY` from the scripts in this lambda folder (note, this cannot be accessed from React components like `StatsBoxGrid.js`). 

We should update `src/lambda/txs.js` and `src/lambda/etherscan.js` with the new endpoints, fetching all the data at once. This allows you to access the process env variable to access the API key in the URL. Note we should not be refetching the data with `onClick` or when toggling between the ranges (especially for node count and tx count since I don't believe these update more than once a day). @samajammin Let me know if you disagree, but I we've been pretty minimalistic with the network calls, and I thought we were trying to keep it that way.

The data can then be passed back to the function that called that lambda function to begin with, which is located in the `useEffect` of `StatsBoxGrid.js`.  If you check out the `fetchNodes` function, line 366 is where this call is made:
```js
          const data = await getData("/.netlify/functions/etherscan")
```
This data can then be parsed into the format you need for the charts. Then store this in the state using a `useState` hook. Separate state variables can be used to differentiate which timeline is selected. Make sure to keep the existing UI function (values being displayed, etc). Hope this helps!

---
Also, noted another warning while looking at this:
`Warning: Each child in a list should have a unique "key" prop.`

Line 573, add the index as a prop named `key`, like this:
```js
{types.map((type, idx) => (
          <ButtonToggle
            active={transactionsActive === type}
            onClick={() => {
              setTransactionsActive(type)
            }}
            key={idx}
          >
```

**Comment by Andrew-Sofos** [comment]
@wackerow I followed your advise and deactivated the etherscan API key and changed it with another that I don't mind getting exposed (not personal email). I also added the key values as you proposed and the error got fixed. I am not really sure if this is the case as far as calling undefined is concerned because I have done a deploy displaying only coingecko data(one of my multiple tries) in which  I was still getting those errors. Either way please let me know when you update the netlify lambda functions to return all the data from the beginning of ethereum (except from txs and etherscan js I think that also the defipulse needs an update). When you achieve this I will be able with just a for loop inside the function to get the appropriate data for the 30d or for all. In the mean while I will try to copy exactly the mock-up that is proposed as far as colours and buttons are concerned.

**Comment by minimalsm** [comment]
Draft PR for this: #3354
Currently blocked by: #3538
